packages/core/src/browser/common-frontend-contribution.ts

```diff
@@ -420,6 +414,36 @@ export class CommonFrontendContribution implements FrontendApplicationContributi
         }
     }
 
+    protected handlePreferenceChange(e: PreferenceChangeEvent<CoreConfiguration>, app: FrontendApplication): void {
+        switch (e.preferenceName) {
+            case 'workbench.editor.highlightModifiedTabs': {
+                this.updateStyles();
+                break;
+            }
+            case 'workbench.colorTheme':
+            case 'workbench.iconTheme': {
+                this.updateThemeFromPreference(e.preferenceName);
+                break;
+            }
+            case 'window.menuBarVisibility': {
+                const { newValue } = e;
+                const mainMenuId = 'main-menu';
+                if (newValue === 'compact') {
+                    app.shell.leftPanelHandler.addTopMenu({
```

Comment:
@ChayaLau we have access to `shell` here, why do we need to pass `app: FrontendApplication`?

```suggestion
                    this.shell.leftPanelHandler.addTopMenu({
```